mmengine/runner/runner.py

```diff
@@ -2080,14 +2089,28 @@ def save_checkpoint(
         if hasattr(self.train_dataloader.dataset, 'metainfo'):
             meta.update(dataset_meta=self.train_dataloader.dataset.metainfo)
 
-        if is_model_wrapper(self.model):
-            model = self.model.module
+        # models with this attribute are FSDP models, which has special logic
+        # for `state_dict()` and should run across all processes
+        if hasattr(self.model, '_mm_fsdp_state_dict'):
+            state_dict = self.model._mm_fsdp_state_dict()
+            # after gathering state_dict, other processes should stop to avoid
+            # multiple saves of a checkpoint
+            if not is_main_process():
+                return
         else:
-            model = self.model
+            # other model should only run in main process
+            if not is_main_process():
+                return
+            if is_model_wrapper(self.model):
+                model = self.model.module
+                state_dict = weights_to_cpu(get_state_dict(model))
+            else:
+                model = self.model
+                state_dict = weights_to_cpu(get_state_dict(model))
 
         checkpoint = {
             'meta': meta,
-            'state_dict': weights_to_cpu(get_state_dict(model)),
+            'state_dict': state_dict,
             'message_hub': self.message_hub.state_dict()
         }
         # save optimizer state dict to checkpoint
```

Comment:
I'm afraid to leave a comment because I don't know if your work is over. It seems that the optimizer state processing for the fsdp model is necessary.  I marked the suggestion on the line closest to where I can leave a comment.

Refer to the [official PyTorch documentation](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.full_optim_state_dict).


https://github.com/open-mmlab/mmengine/blob/24c9627f8ada684a00625296c798c513690256cf/mmengine/runner/runner.py#L2119

```suggestion
        # save optimizer state dict to checkpoint
        if save_optimizer:
            if isinstance(self.optim_wrapper, OptimWrapper):
                checkpoint[
                    'optimizer'] = FullyShardedDataParallel.full_optim_state_dict(
                        self.model, self.optim_wrapper) if isinstance(
                            self.model, FullyShardedDataParallel
                        ) else self.optim_wrapper.state_dict()
```

Reply:
```
import os

import torch
from torch import distributed as dist
from torchvision import models

from torch.distributed.fsdp import FullyShardedDataParallel

os.environ['MASTER_ADDR'] = '127.0.0.1'
os.environ['MASTER_PORT'] = '29500'

dist.init_process_group("nccl", rank=0, world_size=1)

sharded_module = FullyShardedDataParallel(models.resnet50().cuda())

optim_sharded = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
x = sharded_module(torch.randn(2, 3, 32, 32))
loss = x.sum()
loss.backward()
optim_sharded.step()
```

```
>>> print(optim_sharded.state_dict())

{'state': {0: {'step': tensor(1.),
   'exp_avg': tensor([ 2.1422e+06,  4.0932e+05, -1.8306e+06,  ...,  2.0000e-01,
            2.0000e-01,  2.0000e-01], device='cuda:0'),
   'exp_avg_sq': tensor([4.5889e+11, 1.6755e+10, 3.3510e+11,  ..., 4.0000e-03, 4.0000e-03,
           4.0000e-03], device='cuda:0')}},
 'param_groups': [{'lr': 0.0001,
   'betas': (0.9, 0.999),
   'eps': 1e-08,
   'weight_decay': 0,
   'amsgrad': False,
   'maximize': False,
   'foreach': None,
   'capturable': False,
   'params': [0]}]}
```


```
>>> print(FullyShardedDataParallel.full_optim_state_dict(sharded_module, optim_sharded))

{'state': {'conv1.weight': {'exp_avg': tensor([[[[ 2142163.7500,   409323.5625, -1830577.0000,  ...,
              -1085212.7500,  -540307.7500,  1141752.7500],
             [ 2206235.2500, -1877479.0000,  1508397.2500,  ...,
                 89149.3750,  1560391.5000, -1869549.6250],
             [  -38267.7266, -1264628.6250,   -28106.8125,  ...,
                846319.6875, -1516420.6250,  -241758.6562],
             ...,
             [ 1362598.3750,  1829121.2500,  -716642.8125,  ...,
               1271945.8750, -2028220.0000, -1917505.2500],
             [-3755941.7500,   183017.6719,  2154106.2500,  ...,
               2930278.5000, -1245177.1250,  1189422.0000],
             [-1057410.6250,   987793.0000,   593515.1875,  ...,
              -2091312.8750,  -645928.4375,  2445835.7500]],
   
            [[ 1392815.5000,  2420500.5000,  -555766.1875,  ...,
               -633053.6875, -1195849.8750,  -525232.3125],
             [ -864372.8125,    92030.6016,  1542001.2500,  ...,
               2668409.2500, -1473231.8750,  1607459.8750],
             [-2669487.7500,  1017705.4375,  -486886.3438,  ...,
              -1767402.3750,   342614.2500,  4992670.0000],
             ...,
             [ 1064034.6250,  1122348.6250,  1053224.7500,  ...,
                761064.6875,  2284715.7500,  -519216.5938],
             [  383956.0938,  2301994.2500,  -805603.3125,  ...,
               -475890.3125, -2511453.7500, -1123661.7500],
             [ 2440259.5000, -3910662.7500,  -137089.1719,  ...,
              -1700013.3750, -1619611.3750,  -154703.2031]],
   
            [[ -507806.5625,  4081534.7500,   488064.5938,  ...,
                290962.3125,  -965792.3125, -3780304.5000],
             [-1512316.3750, -1566826.8750,  -806461.1875,  ...,
               -276099.0000,  -601556.7500, -3141948.0000],
             [   58800.3242, -3001410.2500,   134112.8750,  ...,
              -1323271.3750,   315482.8125,  -211882.2031],
             ...,
             [ 1523693.3750, -1605059.6250, -2498034.2500,  ...,
               -545667.9375, -2481254.7500,  -575532.4375],
             [-1371176.2500,  -420826.5625, -3567106.7500,  ...,
                348924.1562,   741612.8125,  2181967.5000],
             [  783705.5000,  1125062.3750,   460346.4062,  ...,
                983578.1875,  3038088.7500, -2070376.0000]]],
   
   
           [[[ 1273919.8750,  2697826.2500, -2212009.5000,  ...,
               -469356.3438,   290964.1875,  1549508.2500],
             [ -454629.4062,  1475129.3750,  -786566.1250,  ...,
               1893217.0000,  -359167.7188, -2595191.7500],
             [  -50319.1250,  1031181.8125,  -235231.3594,  ...,
              -2802722.2500,  -366752.0000,   913604.8125],
             ...,
             [-1086281.2500, -1582115.2500,   713219.6875,  ...,
               1654504.7500,  -837480.9375, -1101753.8750],
             [-3507327.2500,  2925520.2500,  2019418.3750,  ...,
               2115340.0000, -2511322.7500,   795400.8750],
             [-2348204.7500, -1722527.0000,  -308280.4688,  ...,
               2022351.3750, -2779469.7500,  -774571.6250]],
   
            [[-1806762.8750,  -186247.2344,   229765.4531,  ...,
              -1469255.0000,  -316652.1875,   530165.8125],
             [ -174926.6562, -1770816.0000,  3064877.0000,  ...,
                155592.4375,   895537.5000,  -293729.8750],
             [ -125938.9766,  -718433.6250,   -42770.3516,  ...,
                934706.1875,  1211416.2500, -1910212.3750],
             ...,
             [ -388171.5938,  -532824.5625,   -70409.6875,  ...,
               2410227.5000,  -303015.2188,  2284995.5000],
             [ 2931384.0000, -1371136.0000, -2880545.0000,  ...,
              -2092458.8750,  -721014.5000, -1690693.6250],
             [ -235694.5000,   853856.1875,  1019186.6250,  ...,
              -1660303.3750,    35478.5508,   584892.7500]],
   
            [[-1569594.3750,  -179152.4531,  1171321.5000,  ...,
               2765002.2500,  -931139.4375,  -828912.8125],
             [  243534.9062,    38038.8008,  1715803.8750,  ...,
               -287563.7188, -2390639.0000,  1788383.8750],
             [ -244658.5000,   156073.7812,  2262097.7500,  ...,
              -3230464.7500, -4350023.5000,  1619839.2500],
             ...,
             [ 2887606.5000,   975075.6250,  1424575.8750,  ...,
                897419.1875,  -864848.0000, -1804312.2500],
             [ 2519565.7500,   629945.1250,    85953.3828,  ...,
               1097585.8750,  -275126.8438, -1216899.6250],
             [ 2257656.2500,  1821216.8750,  -585212.9375,  ...,
               2708101.2500,  -133588.5781, -1439253.2500]]],
   
   
           [[[   37277.0117,  -371206.9688,  -115210.6484,  ...,
               -688240.8750,    -5590.4688, -1574536.6250],
             [  102378.8047,   731538.3750,   209921.4062,  ...,
                213695.6094, -1256450.1250, -2296625.7500],
             [-1686580.6250,  1142392.3750,   320835.3125,  ...,
                789414.7500, -2383125.2500,  -771578.3125],
             ...,
             [-1384804.8750,   795849.6875,   704002.0000,  ...,
              -1233113.3750,   907749.8125,  1770965.3750],
             [-2340262.7500,   385919.5000,   548647.4375,  ...,
               1137807.6250,   542073.1875,  -928521.8125],
             [  -82526.3516,  1150563.8750, -1111282.3750,  ...,
              -1596874.7500,   413789.4375,  -619081.0000]],
   
            [[-1181462.3750,  1018129.3125,    43928.1016,  ...,
                633411.1250, -1365366.8750,  -877781.1250],
             [ -935033.3125,    91189.7891,   931430.5000,  ...,
               2892271.5000,  -225138.1562,  -490295.5000],
             [  153552.0312,  -423469.9062, -1120303.5000,  ...,
              -1003693.1875, -1193527.8750,    24830.0762],
             ...,
             [-2829128.0000,  1729030.6250, -1335298.3750,  ...,
               -958581.9375, -2451720.2500,  -454136.2500],
             [ 3343178.5000, -1013952.6875,   480299.5000,  ...,
               3126308.7500, -1621342.3750,  -234972.4531],
             [ -620935.2500,   224114.5312,  -566348.9375,  ...,
              -2588279.5000,  2158152.5000,  -948973.5000]],
   
            [[-1049963.3750,  3651829.7500,   381169.6250,  ...,
              -1583964.2500,  1449884.0000,   -20235.5254],
             [  625819.6250,  1325850.3750,  -965119.4375,  ...,
               1739562.0000,    10387.5625,   234377.0781],
             [ 1480881.8750,   466732.5000,  -873906.1250,  ...,
              -2116629.7500,  -326607.9688, -1498859.3750],

```